cdap-app-fabric/src/main/java/io/cdap/cdap/internal/sysapp/SystemAppEnableExecutor.java

```diff
@@ -0,0 +1,141 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.sysapp;
+
+import com.google.common.base.Preconditions;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.api.artifact.ArtifactSummary;
+import io.cdap.cdap.api.dataset.DatasetManagementException;
+import io.cdap.cdap.api.retry.RetryableException;
+import io.cdap.cdap.app.program.ProgramDescriptor;
+import io.cdap.cdap.common.ApplicationNotFoundException;
+import io.cdap.cdap.common.ConflictException;
+import io.cdap.cdap.common.InvalidArtifactException;
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.internal.app.deploy.pipeline.ApplicationWithPrograms;
+import io.cdap.cdap.internal.app.services.ApplicationLifecycleService;
+import io.cdap.cdap.internal.app.services.ProgramLifecycleService;
+import io.cdap.cdap.internal.sysapp.SystemAppStep.Arguments;
+import io.cdap.cdap.proto.ProgramStatus;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.KerberosPrincipalId;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.cdap.security.spi.authorization.UnauthorizedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Deploys an application with given arguments. Also runs all programs corresponding to the
+ * application.
+ */
+public class SystemAppEnableExecutor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SystemAppEnableExecutor.class);
+  private static final Gson GSON = new Gson();
+  private final ApplicationLifecycleService appLifecycleService;
+  private final ProgramLifecycleService programLifecycleService;
+
+  @Inject
+  SystemAppEnableExecutor(ApplicationLifecycleService appLifecycleService,
+      ProgramLifecycleService programLifecycleService) {
+    this.appLifecycleService = appLifecycleService;
+    this.programLifecycleService = programLifecycleService;
+  }
+
+  public void deployAppAndStartPrograms(Arguments arguments) {
+    try {
+      // TODO(CDAP-16243): Make deploy app and start programs idempotent to not have transactional issues.
+      ApplicationWithPrograms appDetail = retryableDeploySystemApp(arguments);
+      for (ProgramDescriptor program : appDetail.getPrograms()) {
+        startProgram(program.getProgramId());
+      }
+    } catch (Exception ex) {
+      LOG.error("Failed to deploy app {} and start its program with exception {}.",
+          arguments.getId(), ex);
+
+    }
+  }
+
+  private ApplicationWithPrograms retryableDeploySystemApp(Arguments arguments) throws Exception {
+    return Retries.callWithRetries(() -> deploySystemApp(arguments),
```

Comment:
It looks like nothing is logged if there are failures deploying the app. We need to log errors if they happen, though we also don't want to flood the logs with error messages if it is failing constantly.

You can use the Loggers utility class to create a logger that limits how often it actually logs.

Reply:
Added logging such that it logs in debug mode if error is retryable exception else it will anyways log in catch below.